sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -10,3 +10,175 @@
+  AZ_PRECONDITION_VALID_SPAN(id_scope, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(registration_id, 1, false);
+
+  client->_internal.global_device_endpoint = global_device_endpoint;
+  client->_internal.id_scope = id_scope;
+  client->_internal.registration_id = registration_id;
+
+  if (options != NULL)
+  {
+    client->_internal.options.user_agent = options->user_agent;
+  }
+  else
+  {
+    client->_internal.options = az_iot_provisioning_client_options_default();
+  }
+
+  return AZ_OK;
+}
+
+// <id_scope>/registrations/<registration_id>/api-version=<service_version>
+AZ_NODISCARD az_result az_iot_provisioning_client_user_name_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_user_name,
+    az_span* out_mqtt_user_name)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+
+  const az_span* const user_agent = &(client->_internal.options.user_agent);
+
+  AZ_RETURN_IF_FAILED(az_span_copy(mqtt_user_name, client->_internal.id_scope, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_user_name, str_registrations, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, client->_internal.registration_id, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, provisioning_service_api_version, &mqtt_user_name));
+
+  if (az_span_length(*user_agent) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_user_name, '&', &mqtt_user_name));
+    AZ_RETURN_IF_FAILED(az_span_append(mqtt_user_name, *user_agent, &mqtt_user_name));
+  }
+
+  *out_mqtt_user_name = mqtt_user_name;
+  return AZ_OK;
+}
+
+// <registration_id>
+AZ_NODISCARD az_result az_iot_provisioning_client_id_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_client_id,
+    az_span* out_mqtt_client_id)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_client_id, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_client_id);
+
+  AZ_RETURN_IF_FAILED(
+      az_span_copy(mqtt_client_id, client->_internal.registration_id, &mqtt_client_id));
+
+  *out_mqtt_client_id = mqtt_client_id;
+  return AZ_OK;
+}
+
+// $dps/registrations/res/#
+AZ_NODISCARD az_result az_iot_provisioning_client_register_subscribe_topic_filter_get(
+    az_iot_provisioning_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter)
+{
+  UNUSED(client);
+
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+
+  AZ_RETURN_IF_FAILED(az_span_copy(mqtt_topic_filter, str_dps, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, str_registrations, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic_filter, str_res, &mqtt_topic_filter));
+  AZ_RETURN_IF_FAILED(az_span_append_uint8(mqtt_topic_filter, '#', &mqtt_topic_filter));
```

Comment:
should we build this const or have one?

Reply:
Not sure what you mean. This is a const, compiler will take care of it.